### DISCUSSION TITLE
creating a multi-page TIF with different compression and bit depth on each page

### DISCUSSION THREAD

**Discussion by ghosttie** [dlemstra/Magick.NET#997] [Help]
I have images arriving from a TWAIN scanner, I want to set the bit depth and compression on each page appropriately depending on the number of colors in the image, and save the result as a multi-page TIF.

First I create a MagickImageCollection:

```csharp
var tif = new MagickImageCollection();
```
then when an image arrives from the scanner I create a MagickImage from it, get the total number of colors, set the bit depth and compression, and add it to the MagickImageCollection:

```csharp
using (var stream = e.GetNativeImageStream()) {
	var image = new MagickImage(stream);

	var colors = image.TotalColors;
	if (colors <= 2) {
		image.Settings.Depth = 1;
		image.Settings.Compression = CompressionMethod.Group4;
	} else if (colors <= 8) {
		image.Settings.Depth = 4;
		image.Settings.Compression = CompressionMethod.LZW;
	} else if (colors <= 256) {
		image.Settings.Depth = 8;
		image.Settings.Compression = CompressionMethod.LZW;
	} else if (colors <= 65536) {
		image.Settings.Depth = 16;
		image.Settings.Compression = CompressionMethod.LZW;
	} else if (colors <= 16777216) {
		image.Settings.Depth = 24;
		image.Settings.Compression = CompressionMethod.JPEG;
	} else {
		image.SetCompression(CompressionMethod.JPEG);
	}

	image.Settings.Format = MagickFormat.Tiff;

	tif.Add(image);
}
```

and then when the user is finished scanning I write the MagickImageCollection out as a TIF:

```csharp
var tifStream = new MemoryStream();
tif.Write(tifStream, MagickFormat.Tiff);
```

The problems that I have are:
1. All of the pages appear to use the bit depth and compression from the first page, so if I scan in a b&w page and a 24bit color page, the 24bit color page ends up 1bpp with Group4 compression.
2. If I only scan one page that has a few thousand colors it should end up with a bit depth of 16bpp but it ends up with 24bit.

Are these bugs or am I not doing this correctly?

(I'm using Magick.NET-Q8-AnyCPU v8.0.1 and Windows 10)

EDIT:

I tried to use this instead but it doesn't change the compression or bit depth at all: 

```csharp
image.BitDepth(depth);
image.SetCompression(compressionMethod);
```

**Comment by dlemstra** [comment]
Only the `.Settings` of the first image (page) will be used. If you want to preserve/use the compression of each image when writing a tiff file you will need to do this:

```C#
tif.Write(tifStream, new TiffWriteDefines
{
    PreserveCompression = true,
});
```

This does require you to call `image.SetCompression` on all the images. And it is not possible to force the bit depth with the `.Settings`, you will need to change/reduce the bit depth of the image itself. You can use `image.BitDepth` for that, not sure why this is not working for you/

**Reply by ghosttie** [reply]
Thanks, I've got that all working except that converting the uncompressed 24bpp BMP that I get from the scanner into a 16bpp LZW compressed TIF doesn't work - it ends up being 48bpp instead.

[Here](https://github.com/dlemstra/Magick.NET/files/6964585/test.zip)'s a copy of the uncompressed BMP from the scanner, and here's some code to reproduce the problem:
```csharp
using var bmpStream = File.OpenRead(@"C:\temp\test.bmp");

var image = new MagickImage(bmpStream);

image.Format = MagickFormat.Tiff;
image.BitDepth(16);
image.SetCompression(CompressionMethod.LZW);

using var tifStream = File.Create(@"c:\temp\test.tif");

image.Write(tifStream, new TiffWriteDefines {
	PreserveCompression = true,
});
```

**Reply by dlemstra** [reply]
Are you sure that it is 48 bits per pixel? It looks like you are getting `3*16=48` as the output and that looks like what I would expect. The `BitDepth` in ImageMagick/Magick.NET is the depth per channel instead of the total depth.

**Reply by ghosttie** [reply]
OK I think I understand.

I tried to setting it to 5 and 6 for [High color (15/16-bit)](https://en.wikipedia.org/wiki/Color_depth#High_color_(15/16-bit)) but I got a 2993 color image when High color is supposed to support 32,768 or 65,536.

**Reply by dlemstra** [reply]
The Q8 version of Magick.NET will internally only use 8 bits instead of 16 bits. If you need that bit depth it might be a good idea to use the Q16 version instead.
